src/nbl/builtin/utils.cmake

```diff
@@ -0,0 +1,91 @@
+macro(LIST_BUILTIN_RESOURCE _BUNDLE_NAME_ _LBR_PATH_)
+	set(_OPTIONAL_ALIAS_ "${ARGV2}")
+	
+	list(APPEND ${_BUNDLE_NAME_} "${_LBR_PATH_}")
+	set(${_BUNDLE_NAME_} ${${_BUNDLE_NAME_}} PARENT_SCOPE) # override
+	
+	list(APPEND _LBR_${_BUNDLE_NAME_}_ "${_LBR_PATH_},${_OPTIONAL_ALIAS_}")
+	set(_LBR_${_BUNDLE_NAME_}_ ${_LBR_${_BUNDLE_NAME_}_} PARENT_SCOPE) # override
+endmacro()
+
+function(ADD_CUSTOM_BUILTIN_RESOURCES _TARGET_NAME_ _BUNDLE_NAME_ _BUNDLE_SEARCH_DIRECTORY_ _NAMESPACE_ _OUTPUT_DIRECTORY_)
+	if(NOT DEFINED _Python3_EXECUTABLE)
+		message(FATAL_ERROR "_Python3_EXECUTABLE must be defined - call find_package(Python3 COMPONENTS Interpreter REQUIRED)")
+	endif()
+
+	file(MAKE_DIRECTORY "${_OUTPUT_DIRECTORY_}")
+
+	set(NBL_TEMPLATE_APK_RESOURCES_ARCHIVE_HEADER "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/template/CArchive.h.in")
+	set(NBL_BUILTIN_HEADER_GEN_PY "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/builtinHeaderGen.py")
+	set(NBL_BUILTIN_DATA_GEN_PY "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/builtinDataGen.py")
+	set(NBL_BS_HEADER_FILENAME "builtinResources.h")
+	set(NBL_BS_DATA_SOURCE_FILENAME "builtinResourceData.cpp")
+	
+	string(REPLACE "::" "/" _PATH_PREFIX_ "${_NAMESPACE_}")
+	string(REPLACE "::" "_" _GUARD_SUFFIX_ "${_NAMESPACE_}")
+	string(REGEX REPLACE "^[0-9]+\." "" _GUARD_SUFFIX_ ${_GUARD_SUFFIX_})
+	string(TOUPPER ${_GUARD_SUFFIX_} _GUARD_SUFFIX_)
+	string(MAKE_C_IDENTIFIER ${_GUARD_SUFFIX_} _GUARD_SUFFIX_)
+	
+	set(_ITR_ 0)
+	foreach(X IN LISTS _LBR_${_BUNDLE_NAME_}_)
+		set(_CURRENT_ITEM_ "${X}")
+		string(FIND "${_CURRENT_ITEM_}" "," _FOUND_ REVERSE)
+		
+		string(REGEX REPLACE ",.*" "" _CURRENT_PATH_ "${_CURRENT_ITEM_}")
+		string(REGEX REPLACE ".*," "" _CURRENT_ALIAS_ "${_CURRENT_ITEM_}")
+		
+		if(EXISTS "${_BUNDLE_SEARCH_DIRECTORY_}/${_CURRENT_PATH_}")
+			list(APPEND NBL_DEPENDENCY_FILES "${_BUNDLE_SEARCH_DIRECTORY_}/${_CURRENT_PATH_}")
+			file(SIZE "${_BUNDLE_SEARCH_DIRECTORY_}/${_CURRENT_PATH_}" _FILE_SIZE_)
+			
+			string(APPEND _RESOURCES_INIT_LIST_ "\t\t\t\t\t{\"${_CURRENT_PATH_}\", ${_FILE_SIZE_}, 0xdeadbeefu, ${_ITR_}, nbl::system::IFileArchive::E_ALLOCATOR_TYPE::EAT_NULL},\n") # initializer list
+		else()
+			message(FATAL_ERROR "!") # TODO: set GENERATED property, therefore we could turn some input into output and list it as builtin resource 
+		endif()	
+		
+		math(EXPR _ITR_ "${_ITR_} + 1")
+	endforeach()
+	
+	configure_file("${NBL_TEMPLATE_APK_RESOURCES_ARCHIVE_HEADER}" "${_OUTPUT_DIRECTORY_}/CArchive.h")
+	
+	list(APPEND NBL_DEPENDENCY_FILES "${NBL_BUILTIN_HEADER_GEN_PY}")
+	list(APPEND NBL_DEPENDENCY_FILES "${NBL_BUILTIN_DATA_GEN_PY}")
+
+	set(NBL_RESOURCES_LIST_FILE "${_OUTPUT_DIRECTORY_}/resources.txt")
+
+	string(REPLACE ";" "," RESOURCES_ARGS "${${_BUNDLE_NAME_}}")
+	file(WRITE "${NBL_RESOURCES_LIST_FILE}" "${RESOURCES_ARGS}")
+
+	set(NBL_BUILTIN_RESOURCES_HEADER "${_OUTPUT_DIRECTORY_}/${NBL_BS_HEADER_FILENAME}")
+	set(NBL_BUILTIN_RESOURCE_DATA_SOURCE "${_OUTPUT_DIRECTORY_}/${NBL_BS_DATA_SOURCE_FILENAME}")
```

Comment:
you need two separate output directory, one for headers and one for sources

also the configured headers need to INSTALL

Reply:
resolved in following commit

Reply:
can we make the generated headers install for all targets?

Reply:
as long as we talk about a target that has anything to do with Nabla we could define install rules in the util. The problem we is the util is meant to be included by external projects as well, so for now we use `nbl_install_builtin_resources ` defined here https://github.com/Devsh-Graphics-Programming/Nabla/blob/newBuiltinResources/cmake/common.cmake#L344